src/realm/sync/client_syncsocket.hpp

```diff
@@ -0,0 +1,227 @@
+/// connecting to the server via a WebSocket connection.
+class SyncSocketInterface {
+public:
+    /// The Timer object used to track a timer that was started on the event
+    /// loop.
+    ///
+    /// This object provides a cancel() mechanism to cancel the timer. The
+    /// handler function for this timer must be called with a Status of
+    /// realm::util::error::operation_aborted error code if the timer is
+    /// canceled.
+    ///
+    /// Custom event loop implementations will need to create a subclass of
+    /// Timer that provides access to the underlying implementation to cancel
+    /// the timer.
+    struct Timer {
+        /// Cancels the timer and destroys the timer instance.
+        virtual ~Timer() = default;
+        /// Cancel the timer immediately. Does nothing if the timer has
+        /// already expired or been cancelled.
+        virtual void cancel() = 0;
+    };
+
+    /// The event loop implementation must ensure the event loop is stopped and
+    /// flushed when the object is destroyed. If the event loop is processed by
+    /// a thread, the thread must be joined as part of this operation.
+    virtual ~SyncSocketInterface() = default;
+
+    /// Create a new websocket pointed to the server indicated by endpoint and
+    /// connect to the server. Any events that occur during the execution of the
+    /// websocket will call directly to the handlers provided by the observer.
+    /// The WebSocketObserver guarantees that the WebSocket object will be
+    /// closed/destroyed before the observer is terminated/destroyed.
+    virtual std::unique_ptr<WebSocketInterface> connect(WebSocketObserver* observer, Endpoint&& endpoint) = 0;
+
+    /// Submit a handler function to be executed by the event loop (thread).
+    ///
+    /// Register the specified handler function to be queued on the event loop
+    /// for immediate asynchronous execution. The specified handler will be
+    /// executed by an expression on the form `handler()`. If the the handler
+    /// object is movable, it will never be copied. Otherwise, it will be
+    /// copied as necessary.
+    ///
+    /// This function is thread-safe and can be called by any thread. It can
+    /// also be called from other post() handler function.
+    ///
+    /// The handler will never be called as part of the execution of post(). If
+    /// post() is called on a thread separate from the event loop, the handler
+    /// may be called before post() returns.
+    ///
+    /// Handler functions added through post() must be executed in the order
+    /// they are added. More precisely, if post() is called twice to add two
+    /// handlers, A and B, and the execution of post(A) ends before the
+    /// beginning of the execution of post(B), then A is guaranteed to execute
+    /// before B.
+    ///
+    /// @param handler The handler function to be queued on the event loop.
+    virtual void post(util::UniqueFunction<void(Status)>&& handler) = 0;
```

Comment:
we could create an alias for `util::UniqueFunction<void(Status)>&&`

Reply:
Sure